presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/MetadataQueryOptimizer.java

```diff
@@ -254,27 +272,46 @@ private PlanNode reduce(
                     // min/max ignores null value
                     else if (value.getValue() != null) {
                         Type type = input.getType();
-                        arguments.add(constant(value.getValue(), type));
+                        ConstantExpression constantExpression = constant(value.getValue(), type);
+                        arguments.add(constantExpression);
+                        valueToDomain.putIfAbsent(constantExpression, domain);
                     }
                 }
                 inputColumnValuesBuilder.put(input, arguments.build());
+                inputValueToDomainBuilder.put(input, valueToDomain);
             }
             Map<VariableReferenceExpression, List<RowExpression>> inputColumnValues = inputColumnValuesBuilder.build();
+            Map<VariableReferenceExpression, Map<RowExpression, TupleDomain<ColumnHandle>>> inputValueToDomain = inputValueToDomainBuilder.build();
 
             Assignments.Builder assignmentsBuilder = Assignments.builder();
             for (VariableReferenceExpression outputVariable : node.getOutputVariables()) {
                 Aggregation aggregation = node.getAggregations().get(outputVariable);
-                assignmentsBuilder.put(
-                        outputVariable,
-                        evaluateMinMax(
-                                metadata.getFunctionAndTypeManager().getFunctionMetadata(node.getAggregations().get(outputVariable).getFunctionHandle()),
-                                inputColumnValues.get(getOnlyElement(aggregation.getArguments()))));
+                RowExpression inputVariable = getOnlyElement(aggregation.getArguments());
+                RowExpression result = evaluateMinMax(
+                        metadata.getFunctionAndTypeManager().getFunctionMetadata(node.getAggregations().get(outputVariable).getFunctionHandle()),
+                        inputColumnValues.get(inputVariable));
+                assignmentsBuilder.put(outputVariable, result);
+
+                // Disable the rewrite if the metadata of the resulting partition cannot be trusted.
+                TupleDomain<ColumnHandle> tupleDomain = inputValueToDomain.get(inputVariable).get(result);
+                if (!canStatisticsBeTrusted(table, tupleDomain)) {
+                    return context.defaultRewrite(node);
+                }
             }
             Assignments assignments = assignmentsBuilder.build();
             ValuesNode valuesNode = new ValuesNode(idAllocator.getNextId(), node.getOutputVariables(), ImmutableList.of(new ArrayList<>(assignments.getExpressions())));
             return new ProjectNode(idAllocator.getNextId(), valuesNode, assignments, LOCAL);
         }
 
+        private boolean canStatisticsBeTrusted(TableHandle table, TupleDomain<ColumnHandle> tupleDomain)
```

Comment:
nit: how about `isPositiveRowCount`? I feel this is also not a great name, but "trusted" is a bit ambiguous.

Reply:
Good idea. done.